scripts/Stretch.py

```diff
@@ -0,0 +1,419 @@
+        self.step_z = _getValue(line, "Z", self.step_z)
+        self.step_e = _getValue(line, "E", self.step_e)
+        self.step_f = _getValue(line, "F", self.step_f)
+        return
+
+    def copyPosFrom(self, step):
+        """
+        Copies positions of step into self
+        """
+        self.step_x = step.step_x
+        self.step_y = step.step_y
+        self.step_z = step.step_z
+        self.step_e = step.step_e
+        self.step_f = step.step_f
+        self.comment = step.comment
+        return
+
+
+# Execution part of the stretch plugin
+class Stretcher():
+    """
+    Execution part of the stretch algorithm
+    """
+    def __init__(self, line_width, stretch):
+        self.line_width = line_width
+        self.stretch = stretch
+        self.outpos = GCodeStep(0)
+        self.vd1 = np.empty((0, 2)) # Start points of segments
+                                    # of already deposited material for current layer
+        self.vd2 = np.empty((0, 2)) # End points of segments
+                                    # of already deposited material for current layer
+
+    def execute(self, data):
+        """
+        Computes the new X and Y coordinates of all g-code steps
+        """
+        Logger.log("d", "Post stretch with line width=" + str(self.line_width)
+                   + "mm and stretch=" + str(self.stretch)+ "mm")
+        retdata = []
+        layer_steps = []
+        current = GCodeStep(0)
+        layer_z = 0.
+        for layer in data:
+            lines = layer.rstrip("\n").split("\n")
+            for line in lines:
+                current.comment = ""
+                if line.find(";") >= 0:
+                    current.comment = line[line.find(";"):]
+                if _getValue(line, "G") == 0:
+                    current.readStep(line)
+                    onestep = GCodeStep(0)
+                    onestep.copyPosFrom(current)
+                elif _getValue(line, "G") == 1:
+                    current.readStep(line)
+                    onestep = GCodeStep(1)
+                    onestep.copyPosFrom(current)
+                elif _getValue(line, "G") == 92:
+                    current.readStep(line)
+                    onestep = GCodeStep(-1)
+                    onestep.copyPosFrom(current)
+                else:
+                    onestep = GCodeStep(-1)
+                    onestep.copyPosFrom(current)
+                    onestep.comment = line
+                if current.step_z != layer_z:
+                    Logger.log("d", "Layer Z " + "{:.3f}".format(layer_z)
+                               + " " + str(len(layer_steps)) + " steps")
+                    if len(layer_steps):
+                        retdata.append(self.processLayer(layer_steps))
+                    layer_steps = []
+                    layer_z = current.step_z
+                layer_steps.append(onestep)
+        if len(layer_steps):
+            retdata.append(self.processLayer(layer_steps))
```

Comment:
I agree with you, this is what I should do, but I noticed that the provided array merges sometimes two consecutive layers. For example, with https://www.thingiverse.com/thing:301742 (bottom part) using 0.6mm nozzle and default settings on Ultimaker 2+, layers 139 and 140 are merged.
To check it, I just created a very simple script:

```
    def execute(self, data):
        for layer_number, layer in enumerate(data):
            data[layer_number] = ";PPLayer" + str(layer_number) + "\n" + data[layer_number];
        return data;
```

The first three items of the data array contain comments, following items follow correctly the layer changes... until layer 139.

And then, from layer 141 to the end, separation is correct.

Maybe a Cura bug 5 layers before the last one?